p2p/src/collection/mod.rs

```diff
@@ -0,0 +1,70 @@
+//! An [Originator] sends messages out to a set of [Endpoint]s and collects responses.
+//!
+//! The originator's [Collector] is used to distribute and collect responses.
+//! The endpoint's [Collector] is used to receive and respond to messages.
+
+use bytes::Bytes;
+use commonware_cryptography::PublicKey;
+use futures::channel::oneshot;
+use std::{collections::HashMap, fmt::Debug, future::Future};
+
+pub mod actor;
+
+/// Trait for an item that has an associated, unique ID.
+pub trait Idable: Clone + Send + 'static {
+    type ID: Debug + Send + 'static;
+
+    /// Returns the unique ID of the item.
+    fn id(&self) -> Self::ID;
+}
+
+/// Interface for actor that will disperse messages and collect responses.
+pub trait Collector: Clone + Send + 'static {
+    type Message: Idable + Debug + Send + 'static;
+    type PublicKey: PublicKey;
+
+    /// Sends a message to recipients.
+    ///
+    /// Once a quorum of responses have been collected, the [Originator] will be notified.
+    fn send(
+        &mut self,
+        message: Self::Message,
+        transformer: fn(Self::Message, Self::PublicKey) -> Bytes,
+    ) -> impl Future<Output = ()> + Send;
+
+    /// Peek at the collected responses for a given ID.
+    fn peek(
+        &mut self,
+        id: <Self::Message as Idable>::ID,
+    ) -> impl Future<Output = oneshot::Receiver<HashMap<Self::PublicKey, Bytes>>> + Send;
+
+    /// Cancels a message, dropping all existing and future responses.
+    fn cancel(&mut self, id: <Self::Message as Idable>::ID) -> impl Future<Output = ()> + Send;
+}
+
+/// Interface for the application that originates messages.
+pub trait Originator: Clone + Send + 'static {
+    type PublicKey: PublicKey;
+    type ID: Debug + Send + 'static;
```

Comment:
I think we should use `Array` here for consistency elsewhere in the codebase?

Like: https://github.com/commonwarexyz/monorepo/blob/da9ba1498f7d2906d88b8a3d4f733e267eb4ac2f/resolver/src/p2p/mod.rs#L52-L53